File: static/python/symmetric_binary_tree.py
from binary_tree_prototype import BinaryTreeNode


# @include
def is_symmetric(tree):
    def check_symmetric(subtree_0, subtree_1):
        if not subtree_0 and not subtree_1:
            return True
        elif subtree_0 and subtree_1:
            return (subtree_0.data == subtree_1.data and check_symmetric(
                subtree_0.left, subtree_1.right) and check_symmetric(
                    subtree_0.right, subtree_1.left))
        # One subtree is empty, and the other is not.
        return False

    return not tree or check_symmetric(tree.left, tree.right)
# @exclude


def simple_test():
    symm_tree = BinaryTreeNode()
    assert is_symmetric(symm_tree)
    symm_tree.left = BinaryTreeNode()
    assert not is_symmetric(symm_tree)
    symm_tree.right = BinaryTreeNode()
    assert is_symmetric(symm_tree)
    symm_tree.right.right = BinaryTreeNode()
    assert not is_symmetric(symm_tree)


def main():
    simple_test()
    # Non symmetric tree test.
    #      3
    #    2   5
    #  1    4 6
    non_symm_tree = BinaryTreeNode()
    non_symm_tree.left = BinaryTreeNode()
    non_symm_tree.left.left = BinaryTreeNode()
    non_symm_tree.right = BinaryTreeNode()
    non_symm_tree.right.left = BinaryTreeNode()
    non_symm_tree.right.right = BinaryTreeNode()
    assert not is_symmetric(non_symm_tree)
    print(is_symmetric(non_symm_tree))
    # Symmetric tree test.
    symm_tree = BinaryTreeNode()
    symm_tree.left = BinaryTreeNode()
    symm_tree.right = BinaryTreeNode()
    assert is_symmetric(symm_tree)
    print(is_symmetric(symm_tree))
    # Empty tree test.
    symm_tree = None
    assert is_symmetric(symm_tree)
    print(is_symmetric(symm_tree))


if __name__ == '__main__':
    main()

File: static/solutions/Symmetric_binary_tree.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>

#include "./Binary_tree_prototype.h"

using std::boolalpha;
using std::cout;
using std::endl;
using std::make_unique;
using std::unique_ptr;

bool CheckSymmetric(const unique_ptr<BinaryTreeNode<int>>&,
                    const unique_ptr<BinaryTreeNode<int>>&);

// @include
bool IsSymmetric(const unique_ptr<BinaryTreeNode<int>>& tree) {
  return tree == nullptr || CheckSymmetric(tree->left, tree->right);
}

bool CheckSymmetric(const unique_ptr<BinaryTreeNode<int>>& subtree_0,
                    const unique_ptr<BinaryTreeNode<int>>& subtree_1) {
  if (subtree_0 == nullptr && subtree_1 == nullptr) {
    return true;
  } else if (subtree_0 != nullptr && subtree_1 != nullptr) {
    return subtree_0->data == subtree_1->data &&
           CheckSymmetric(subtree_0->left, subtree_1->right) &&
           CheckSymmetric(subtree_0->right, subtree_1->left);
  }
  // One subtree is empty, and the other is not.
  return false;
}
// @exclude

void SimpleTest() {
  auto symm_tree = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  assert(IsSymmetric(symm_tree));
  symm_tree->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  assert(!IsSymmetric(symm_tree));
  symm_tree->right = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  assert(IsSymmetric(symm_tree));
  symm_tree->right->right =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  assert(!IsSymmetric(symm_tree));
}

int main(int argc, char* argv[]) {
  SimpleTest();
  // Non symmetric tree test.
  //      x
  //    x   x
  //  x    x x
  auto non_symm_tree =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  non_symm_tree->left =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  non_symm_tree->left->left =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  non_symm_tree->right =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  non_symm_tree->right->left =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  non_symm_tree->right->right =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  assert(!IsSymmetric(non_symm_tree));
  cout << boolalpha << IsSymmetric(non_symm_tree) << endl;
  // Symmetric tree test.
  unique_ptr<BinaryTreeNode<int>> symm_tree =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  symm_tree->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  symm_tree->right = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>());
  assert(IsSymmetric(symm_tree) == true);
  cout << boolalpha << IsSymmetric(symm_tree) << endl;
  // Empty tree test.
  symm_tree = nullptr;
  assert(IsSymmetric(symm_tree) == true);
  cout << boolalpha << IsSymmetric(symm_tree) << endl;
  return 0;
}
